pkg/fswatcher/fswatcher.go

```
@@ -64,37 +88,65 @@ type Watcher struct {
+
+	target     string      // target path, only applicable to symlinks
+	targetInfo os.FileInfo // target info, only applicable to symlinks
+}
+
+// Option to configure the Watcher
+type Option func(*Watcher)
+
+// WithInterval sets the interval at which the Watcher checks for changes
+func WithInterval(d time.Duration) Option {
+	return func(w *Watcher) {
+		w.interval = d
+	}
+}
+
 // New creates a new Watcher which watches all trackedFile paths (they do not
 // need to exist yet).
-func New(trackedFiles []string) (*Watcher, error) {
-	watcher, err := fsnotify.NewWatcher()
-	if err != nil {
-		return nil, err
+func New(trackedFiles []string, options ...Option) (*Watcher, error) {
+	// sane default that is configurable via WithInterval depending on the use case
+	defaultInterval := 5 * time.Second
+	if testing.Testing() {
+		// in a testing situation, refresh info much much faster to speed things up
+		defaultInterval = 10 * time.Millisecond
 	}
 
 	w := &Watcher{
-		watcher:              watcher,
-		watchedPathCount:     counter.Counter[string]{},
-		trackedToWatchedPath: map[string]string{},
-		Events:               make(chan Event),
-		Errors:               make(chan error),
-		stop:                 make(chan struct{}),
+		Events:   make(chan Event),
+		Errors:   make(chan error),
+		stop:     make(chan struct{}),
+		interval: defaultInterval,
 	}
 
-	// We add all paths in the constructor avoid the need for additional
-	// synchronization, as the loop goroutine below will call updateWatchedPath
-	// concurrently
+	// make a map of tracked files and assign them all empty state at the start
+	tracked := make(map[string]state, len(trackedFiles))
 	for _, f := range trackedFiles {
-		err := w.updateWatchedPath(f)
-		if err != nil {
-			return nil, err
-		}
+		tracked[f] = state{path: f}
 	}
+	w.tracked = tracked
 
+	for _, option := range options {
+		option(w)
+	}
+
+	if w.interval != 0 {
+		w.ticker = time.Tick(w.interval)
+	}
```

Shouldn't we return an error on `w.interval == 0`? If we don't `w.ticker` is `nil` and we will never attempt to read anything? If that is the intended behavior then we could remove the return `error` from `New`?

**EDIT**: ok after reading the tests it's clearer to me that it is intended: we setup `WithInterval(0)` to call `tick()` from the test in order to control exactly the fs state when the watcher take a look. No objection on the pattern, OTHO not a huge fan of exposing the behavior outside of the package (from the package test we have direct access to the `ticker` / channel).

---

I will try to find a neater pattern to do this only from the tests